engine/engine/content/builtins/scripts/debugger.lua

```diff
@@ -25,9 +25,162 @@ local function onexit(code, close)
   msg.post("@system:", "exit", {code = code})
 end
 
+-- Lightweight hook layer
+
+local mobdebug_hook
+local mobdebug_on
+local breakpoints = {}
+local coroutines = {}; setmetatable(coroutines, {__mode = "k"}) -- "weak" keys
+local continue_execution = true
+local exit_current_lua_stack = true
+local last_co = nil
+local stack_after_suspend = 0
+
+local function set_breakpoint(line, file)
+  -- Mobdebug save breakpoints in [line][file] table
+  -- but for our usecase it's better to have [file][line] table
+  if not breakpoints[file] then
+    breakpoints[file] = {}
+  end
+  breakpoints[file][line] = true
+end
+
+local function restore_mobdebug_hook(co)
+  -- Restore mobdebug hook for line-by-line hook
+  if co then
+    debug.sethook(co, mobdebug_hook, "lcr")
+  else
+    debug.sethook(mobdebug_hook, "lcr")
+  end
+end
+
+local function remove_breakpoint(line, file)
+  if breakpoints[file] then
+    breakpoints[file][line] = nil
+  end
+  if breakpoints[file] and not next(breakpoints[file]) then
+    breakpoints[file] = nil
+  end
+end
+
+-- This hook fires only on Lua functions calls and checks if this function has lines
+-- we have breakpoints on
+local function lightweight_hook()
+  local caller = debug.getinfo(2,"S")
+  local file = mobdebug.get_filename(caller.source)
+  mobdebug.get_server_data(file)
+  local brk = breakpoints[file]
+  local co = coroutine.running()
+  if co then
+    -- If it's the first call after the [C] call - save function metricks
+    if caller.lastlinedefined ~= -1 and coroutines[co] == false then
+        -- Coroutine provide this info only on the first coroutine fn call
+        -- we should save it for later usage
+        coroutines[co] = {
+          ["lines"] = debug.getinfo(2,"L"),
+          ["file"] = file,
+          ["caller"] = caller
+        }
+    end
+    -- Make sure [C] function call was before the first coroutine call
+    if not coroutines[co] and caller.lastlinedefined == -1 then
```

Comment:
What of co is nil here? The code will run but what is the purpose of coroutines[nil] = false in that case? Perhaps also check if `co` is not nil?

Reply:
Oh, never mind. I'm not reading the indentation right... sorry. Ignore my question above!